### ISSUE TITLE
Hang due to hdf5 solution files output

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#961]
I am experiencing a hang which I think is due to hdf5 solution files output. In order to reproduce it quickly, I commented all my task bodies and outputs these files at every iteration. After a few iterations, I can consistently reproduce the hang. 

Due to the fact that I don't use a copy launcher and to current limitations of the control_replication branch, I output one file per rank. I have an index launch task which creates and destroys inside its body a region attached to that rank's file. Then I use READ_WRITE privileges to write to the region attached to the file.

I ran on 2 ranks with `-ll:force_kthreads` on 2 nodes to get these backtraces.

[gdb0.txt](https://github.com/StanfordLegion/legion/files/5468457/gdb0.txt)
[gdb1.txt](https://github.com/StanfordLegion/legion/files/5468458/gdb1.txt)

**Comment by lightsighter** [comment]
Run with `-lg:no_physical_tracing` and report back what happens.

**Comment by bandokihiro** [comment]
Hanged at the same iteration
[gdb0.txt](https://github.com/StanfordLegion/legion/files/5471381/gdb0.txt)
[gdb1.txt](https://github.com/StanfordLegion/legion/files/5471382/gdb1.txt)

For reference my command line was
```
mpiexec -np 2 -ppn 1 /nobackupp2/kbando/LeDGend/build_flame_spherical_commented/exec/solver -logfile logs/log_%.log -hdf5:forcerw -dm:memoize -ll:util_bgwork 0 -ll:cpu 1 -ll:util 2 -ll:dma 2 -ll:csize 2000 -ll:rsize 0 -ll:onuma 0 -ll:ocpu 1 -ll:othr 1 -ll:force_kthreads -lg:no_physical_tracing
```

**Comment by lightsighter** [comment]
Run with `-lg:inorder` next and see if it reproduces. If it does, capture backtraces again.

**Comment by bandokihiro** [comment]
It reproduced.
[gdb0.txt](https://github.com/StanfordLegion/legion/files/5482332/gdb0.txt)
[gdb1.txt](https://github.com/StanfordLegion/legion/files/5482333/gdb1.txt)

**Comment by lightsighter** [comment]
The fact that this program hangs with `-lg:inorder` suggests that this is highly unlikely to be a Legion problem. It's either going to be Realm or HDF5 that are the problem. Those sub-tasks are not running probably because they are depending copies coming from or going to an HDF5 instance. What dependences are you expecting for the sub-tasks (the being launched inside of non-replicated contexts)?

@streichler for visibility since I think this either going to be Realm or HDF5's problem.

**Comment by streichler** [comment]
The test is hanging because tasks are waiting on an allocation in sysmem to complete.  The allocation won't complete because it's a deferred allocation, and is waiting for one of the 18 deferred deallocations in that memory to actually complete.  I didn't look at all 18 of them, but all of the ones I did have user events as preconditions and those user events have not been triggered (not even a deferred trigger).  Passing the ball back to @lightsighter to determine why these deallocations are being held up.

**Comment by lightsighter** [comment]
I need a way to run this myself on sapling if you want me to debug it.

**Comment by lightsighter** [comment]
Remember also, the runtime currently has no protections against doing deferred allocations/deletions in program order at the moment, which means this could be a resource deadlock and it will be the mapper's responsibility to fix it.

**Comment by bandokihiro** [comment]
you can run `/home/bandokihiro/Runs/flame2D/run.sh` from `/home/bandokihiro/Runs/flame2D` on sapling, it will hang after printing 
```
SSPRK3 step from 8e-08 to 8.5e-08 with time-step 5e-09
SSPRK3 step from 8e-08 to 8.5e-08 with time-step 5e-09
```
with the current settings of the run file

**Comment by bandokihiro** [comment]
At some point, I asked if it is possible to explicitly release regions that are no longer in use. At that time, the conclusion was that garbage collection was done lazily and I don't remember if there is a way to explictly tell the mapper to get rid of some specific instances.

**Comment by lightsighter** [comment]
Pull and try again.

I'm still confused about why Realm thinks it needs to do deferred allocations in an HDF memory with a capacity of zero.

**Comment by streichler** [comment]
The deferred allocations were in sysmem, not the HDF5 memory.

**Comment by bandokihiro** [comment]
still hanging

**Comment by lightsighter** [comment]
There are two references keeping the sysmem instances alive. One is definitely the application's the fault. The second is unclear.

First, the mapper is holding `GC_NEVER_PRIORITY` references to every single one of the instances made in the system memory. That keeps them alive forever until the mapper releases them. And before Sean asks: no the runtime does not consider this when doing deferred deletions. Mappers know what they are doing with physical instance allocation and it is not the runtime's responsibility to hold their hand to prevent resource deadlocks.

There is a second reference being held on the instance and I can't tell what it is without more information. It could be that the application is not deleting the associated logical region/field or it could be a runtime bug. I need a binary built with `-DDEBUG_LEGION_GC` and no optimizations (`-O0`) in order to figure out the provenance though.

This is a perfect example of where having a small reproducible test that I can run myself would go a long way to shortening these debug cycles so I don't have to keep waiting for you to rebuild these things. This is a completely deterministic thing that is happening on one node. There should be no need for multiple nodes or a full application to reproduce it.

**Comment by lightsighter** [comment]
Just as an example to the previous comment, look at issues #960 and #963 that Manolis filed this week. Both had very simple reproducer programs and as a result we resolved both of his problems in short order.

**Comment by bandokihiro** [comment]
I understand. I will try to make one.

There is a higher cost on my side to write a reproducer since it often takes me a decent amount of time to implement a new piece of code that involves a lot of legion stuff. That is why I always try to see if something can be discovered first with the full application which has a reasonably robust infrastructure for the legion part. I will think of a better way to manage these kind of things.

**Comment by lightsighter** [comment]
Don't bother doing it for this one since I've already dug around enough in the current binary to know what things should look like. Just provide the new binary for now compiled with `-DDEBUG_LEGION_GC` and `-O0`.

**Comment by bandokihiro** [comment]
updated

**Comment by lightsighter** [comment]
It looks like on the runtime side the task contexts are holding a reference to the equivalence set for too long, but I can't see yet which context it is.

Can we switch to the `neweq` branch and trying there?

**Comment by bandokihiro** [comment]
it works with `neweq`. Looking at the issues you mentioned previously, I guess you will want to wait for it to merge into `control_replication` which is I think fine for now by me (increasing `csize` will delay this issue sufficiently for now).

Regarding your previous comment, I am a bit confused as to what I am doing wrong with the mapper. My expectation is that I am using the default mapper behavior for this case and I am not doing anything particular (https://github.com/IhmeGroup/DG-Legion/blob/89a03fc34587859ffa34fb2a338dd51a5a8aaaad/tasks/solution_tasks_impl.h#L4194-L4286). Am I missing something?

**Comment by lightsighter** [comment]
> I guess you will want to wait for it to merge into control_replication which is I think fine by me for now (increasing csize will delay this issue sufficiently for now).

Yes. If it were in `master` I would consider fixing it separately. But `neweq` is likely to merge into `control_replication` sometime this month, so it is not worth taking the time to fix for now.

> My expectation is that I am using the default mapper behavior for this case and I am not doing anything particular 

The default mapper has an implicit assumption that all of regions in the program have the lifetime of the entire program and therefore it will hold onto instances forever. For programs with temporary regions, such as yours, this assumption is faulty, so you need a custom mapper that doesn't make the same assumption. The runtime is helping you out in the `neweq` branch by removing those never gc references after regions are deleted, but in general it is bad form to have the mapper continuing to hold onto physical instances after they are dead. At a minimum the data structures storing instance names in the default mapper are still growing unbounded and there's nothing the runtime can do to stop that. In general, this is why you shouldn't rely on the default mapper for any kind of performance runs or managing any kind programs with resource constrained mappings.

**Comment by bandokihiro** [comment]
Closing this issue since `neweq` has been merged to `control_replication`.
